nixos/modules/services/networking/powerdns.nix

```diff
@@ -4,15 +4,24 @@ with lib;
 
 let
   cfg = config.services.powerdns;
-  configDir = pkgs.writeTextDir "pdns.conf" "${cfg.extraConfig}";
-in {
+  configDir = pkgs.writeTextDir "pdns.conf" (concatStringsSep "\n" (mapAttrsToList (key: value: "${key}=${toStr value}") cfg.settings));
+  toStr = value:
+    if value == true then "yes"
+    else if value == false then "no"
+    else builtins.toString value
+  ;
+in
+{
+  imports = [
+    (mkRemovedOptionModule [ "services" "powerdns" "extraConfig" ] "Use services.powerdns.settings instead.")
+  ];
+
   options = {
     services.powerdns = {
       enable = mkEnableOption "PowerDNS domain name server";
 
-      extraConfig = mkOption {
-        type = types.lines;
-        default = "launch=bind";
+      settings = mkOption {
+        type = with types; attrsOf (oneOf [ bool int str ]);
```

Comment:
The type is bit more complex than this: there options accepting a list of values, for example.
You should be able to reuse the [functions](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/networking/pdns-recursor.nix#L8) I've defined for pdns-recursor, which seems to be using the same syntax.
We could move them to a common file and import them in both modules.

Reply:
ping @Infinisil - there are several pieces of `powerdns` related software that all uses the same configuration format. This format is different enough from existing formats (`ini`) that it doesn't quite match anything we have. Should I add a new `pkgs.formats.pdns` format, or abstract this somewhere else?

Thanks!

Reply:
I think unless we have a clear specification or some indication that two programs use the same format, we shouldn't abstract this. In this case it seems to be simple enough that we can just define the formats inline in the modules themselves. I guess we could add a `types.oneOrMore` at some point though. Fyi I think this should maybe rather be `oneOrMore = type: coercedTo type singleton (listOf type)`, so that merging works between single and multiple elements.

Reply:
Thanks for the answer @Infinisil.